tests/visualization_tests/test_contour.py

```diff
@@ -79,252 +108,443 @@ def fail_objective(_: Trial) -> float:
-        "xaxis4": (param_a_range, param_a_type),
-        "xaxis5": (param_b_range, param_b_type),
-        "xaxis6": (param_c_range, param_c_type),
-        "xaxis7": (param_a_range, param_a_type),
-        "xaxis8": (param_b_range, param_b_type),
-        "xaxis9": (param_c_range, param_c_type),
-        "yaxis": (param_a_range, param_a_type),
-        "yaxis2": (param_a_range, param_a_type),
-        "yaxis3": (param_a_range, param_a_type),
-        "yaxis4": (param_b_range, param_b_type),
-        "yaxis5": (param_b_range, param_b_type),
-        "yaxis6": (param_b_range, param_b_type),
-        "yaxis7": (param_c_range, param_c_type),
-        "yaxis8": (param_c_range, param_c_type),
-        "yaxis9": (param_c_range, param_c_type),
+    info = _get_contour_info(study)
+    params = ["param_a", "param_b", "param_c"]
+    assert info.sorted_params == params
+    assert np.shape(np.asarray(info.sub_plot_infos)) == (3, 3, 3)
+    ranges = {
+        "param_a": (math.pow(10, -6.05), math.pow(10, -4.95)),
+        "param_b": (-0.05, 1.05),
+        "param_c": (-0.05, 1.05),
     }
-
-    for axis, (param_range, param_type) in axis_to_range_and_type.items():
-        assert figure.layout[axis]["range"] == param_range
-        assert figure.layout[axis]["type"] == param_type
-
-
-def test_plot_contour_mixture_category_types() -> None:
+    is_log = {"param_a": True, "param_b": False, "param_c": False}
+    is_cat = {"param_a": False, "param_b": True, "param_c": True}
+    indices: Dict[str, List[Union[str, float]]] = {
+        "param_a": [math.pow(10, -6.05), 1e-6, 1e-5, math.pow(10, -4.95)],
+        "param_b": ["100", "101"],
+        "param_c": ["one", "two"],
+    }
+    values = {"param_a": [1e-6, 1e-5], "param_b": ["101", "100"], "param_c": ["one", "two"]}
+
+    def _check_axis(axis: _AxisInfo, name: str) -> None:
+        assert axis.name == name
+        assert axis.range == ranges[name]
+        assert axis.is_log == is_log[name]
+        assert axis.is_cat == is_cat[name]
+        assert axis.indices == indices[name]
+        assert axis.values == values[name]
+
+    for yi in range(3):
+        for xi in range(3):
+            xaxis = info.sub_plot_infos[yi][xi].xaxis
+            yaxis = info.sub_plot_infos[yi][xi].yaxis
+            x_param = params[xi]
+            y_param = params[yi]
+            _check_axis(xaxis, x_param)
+            _check_axis(yaxis, y_param)
+            z_values = info.sub_plot_infos[yi][xi].z_values
+            if xi == yi:
+                assert z_values == {}
+            else:
+                for i, v in enumerate([0.0, 1.0]):
+                    x_value = xaxis.values[i]
+                    y_value = yaxis.values[i]
+                    assert x_value is not None
+                    assert y_value is not None
+                    xi = xaxis.indices.index(x_value)
+                    yi = yaxis.indices.index(y_value)
+                    assert z_values[(xi, yi)] == v
```

Comment:
Ditto.

Reply:
A direct comparison with NamedTuple would make the test very long, so I will leave it at that here.